Add optimization summary documents

a7ehuo opened the pull request:
Added the following four documents on the summary of the JIT optimizations. These four documents are extracted
from the document written by Vijay - `"Summary of optimizations in the Testarossa JIT"` stored in Notes DB `InsideTRJIT`.

- LocalOptimizationsSummary.md
- GlobalOptimizationsSummary.md
- ControlFlowOptimizationsSummary.md
- LoopOptimizationSummary.md

Authored-By: Vijay Sundaresan <vijaysun@ca.ibm.com>

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

a7ehuo commented:
@vijaysun-omr @hzongaro May I ask you to review these documents? Thank you!
@0xdaryl fyi

hzongaro reviewed (commented):
Thanks for pulling this together, Annabelle!  I haven't finished reviewing yet, but I thought I would submit some of my comments for the first parts.

hzongaro commented:
The commit comment mentions a database that's not accessible to most people in the community.  I would suggest that you mention that to make that clear.

dsouzai commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+
+## Global Copy Propagation (Dead Store Elimination, Dead Structure Removal)
+These three analyses use use/def info to perform the traditional optimizations. The
+analysis is only done for locals (statics/fields omitted). Dead structure removal
```
```suggestion
analysis is only done for locals (statics/fields are omitted). Dead structure removal
```

dsouzai commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+
+## Redundant Monitor Elimination
+Redundant Monitor Elimination uses value numbering to perform nested monitor removal
+(e.g. syncenter on the same object twice followed by two syncexits). Use value
```
```suggestion
(e.g. syncenter on the same object twice followed by two syncexits). It uses value
```

dsouzai commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+physical register the value is placed into). Global register candidates are chosen
+by various optimizer passes like PRE, induction variable detection etc. and appropriate
+priorities are given to the candidates based on the number of references and live
+range (number of blocks). Advantages are that most of the code is platform neutral
```
```suggestion
range (number of blocks). The advantages are that most of the code is platform neutral
```

vijaysun-omr replied:
Suggestion : `most of the code is platform neutral and is` -> `most of the code is platform neutral apart from a few simple evaluators that must be added to a new platform code generator, thus drastically reducing the development cost of adding a powerful global register allocator. It is also `

dsouzai commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+by various optimizer passes like PRE, induction variable detection etc. and appropriate
+priorities are given to the candidates based on the number of references and live
+range (number of blocks). Advantages are that most of the code is platform neutral
+and it is relatively cheap in compile time in comparison to (expected cost of) a
```
```suggestion
and it is relatively cheap in compile time in comparison to (the expected cost of) a
```

dsouzai commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+
+## Live Variables for GC (local and global)
+Live Variables for GC attempts to minimize the number of locals that need to be zero
+initialized upon method entry. Basic idea is to perform liveness analysis and find
```
```suggestion
initialized upon method entry. The basic idea is to perform liveness analysis and find
```

dsouzai commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+out if any locals are live at any [GC point](https://github.com/eclipse/omr/blob/81b79405da6c7c960e611a8b2b12fd5861543330/compiler/il/OMRNode.cpp#L2298)
+that is not dominated by (user defined) stores on all control flow paths. If a local
+is uninitialized at a GC point, explicit zero initialization must be inserted at the method entry.
+Note that this situation is fairly rare (only occurs in some uncommon cases involving jsr).
```
"Involving JSR": do you mean JSR292? JSR by itself just refers to Java Specification Request.

vijaysun-omr replied:
No, a long long time ago in a galaxy far far away, there used to be a bytecode known as `jsr` (jump to sub routine) that was used to implement throw/catch/finally when Java was first released. Because this code essentially caused some abnormal control flow, it could make it appear as though a variable was uninitialized on some code path. The `jsr` bytecode was removed a long time back. 

I believe the main situation in which such zero initializations may be needed in today's JIT compiler may be around some optimization creating such an appearance of uninitialized variables. 

I wonder if we should add the word "reference"  to "local" since the GC cares only about reference locals.

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+Tree Simplification performs simple IL tree transformations like constant
+folding, strength reduction, canonicalizing expressions (e.g. an `add` with a
+constant child will have the constant as the second child), branch/switch
+simplification etc. Float/double simplification support needs to be improved
+(call C libraries to compute the correct answer as per the Java spec).
```
When you say support needs to be improved, are you calling out a shortcoming of the current state? If so, it probably doesn't make sense to have it in a documentation setting, unless you can reword it to describe what the current simplification does and that it has opportunity for further refinement.

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+of subsumed checks and compares. It also performs devirtualization
+of calls, checkcast and array store check elimination. It does not do value
+numbering, instead it works on the assumption that every expression has a
+unique value number (improved effectiveness if expressions having same value
+are commoned by a prior local CSE pass).
```
```suggestion
unique value number; its effectiveness is improved if expressions having same value
are commoned by a prior local CSE pass.
```

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+
+## Array Length Simplification
+In OpenJ9, array lengths are stored in number of bytes. So bound checks require
+a shift right (depending on the array stride) on the array length so that
```
```suggestion
a shift right (depending on the array stride) on the array length so that the
```

vijaysun-omr replied:
This optimization no longer is needed since arraylengths were changed in OpenJ9 to be stored as "number of elements" rather than "number of bytes" 20 years ago :)

vijaysun-omr replied:
To be 100% clear, I am saying this optimization does not exist for the above reason anymore and so it also should not be documented.

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+## Array Length Simplification
+In OpenJ9, array lengths are stored in number of bytes. So bound checks require
+a shift right (depending on the array stride) on the array length so that
+compare can be done with array index. This optimization attempts to eliminate
```
```suggestion
compare can be done with the array index. This optimization attempts to eliminate
```

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+In OpenJ9, array lengths are stored in number of bytes. So bound checks require
+a shift right (depending on the array stride) on the array length so that
+compare can be done with array index. This optimization attempts to eliminate
+`shr` wherever possible (instead comparing the index left shifted based on
```
```suggestion
`shr` wherever possible (instead comparing the index left shifted based on the
```

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+
+## Compaction of Null Explicit Null Checks
+Compaction of Null Explicit Null Checks minimize the cases where an explicit null
+check needs to be performed (null checks are implicit, i.e. no code for doing null
+check, instead we use hardware trap on IA32). Sometimes (say as a result of dead
```
```suggestion
check needs to be performed (null checks are implicit, i.e. by default on IA32, the JIT does not
emit code to perform a null check, instead we use a hardware trap). Sometimes (e.g. as a result of dead
```

dsouzai replied:
Do we have implicit null checks only on IA32?

vijaysun-omr replied:
No, we have them on X86 (32 and 64 bitness). IA32 was just the quaint way of referring to the platform when "64-bit was new".

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+be the same if evaluated later). This optimization attempts to simply delay evaluation
+of expressions and has the effect of reducing register pressure by shortening
+live ranges. It has the effect of removing some IL trees that are simply anchors
+for (already) evaluated expressions and making the trees more compact.
```
```suggestion
for (already) evaluated expressions thus making the trees more compact.
```

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+for (already) evaluated expressions and making the trees more compact.
+
+## Rematerialization
+Rematerialization is performed as a late pass before code generation is to be done.
```
```suggestion
Rematerialization is performed as a late pass before code generation.
```

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+
+## Rematerialization
+Rematerialization is performed as a late pass before code generation is to be done.
+Basically queries the underlying code generator (IA32 etc.) for the max number of
```
```suggestion
The optimization queries the underlying code generator (IA32 etc.) for the max number of
```

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+Rematerialization is performed as a late pass before code generation is to be done.
+Basically queries the underlying code generator (IA32 etc.) for the max number of
+registers available for assignment simultaneously, and then tries to reverse
+commoning in regions of code within a block where number of live expressions could
```
```suggestion
commoning in regions of code within a block where the number of live expressions could
```

dsouzai commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+registers available for assignment simultaneously, and then tries to reverse
+commoning in regions of code within a block where number of live expressions could
+cause spills. Only certain kinds of expressions are considered for rematerialization,
+e.g. address calculations (can often use more complex addressing modes on a chip),
```
```suggestion
e.g. address calculations (as chips can support more complex addressing modes),
```

dsouzai commented on doc/compiler/optimizer/LoopOptimizationSummary.md:
```diff
@@ -0,0 +1,114 @@
+checkcasts and inline guards (where the `this` does not change in the loop)
+are eliminated from loops by placing equivalent tests outside the loop. It
+also eliminates bound checks where the range of values of the index inside
+the loop is discovered (need to be a compile time constant) and checks are
```
```suggestion
the loop is discovered (if it is a compile time constant) and checks are
```

dsouzai commented on doc/compiler/optimizer/LoopOptimizationSummary.md:
```diff
@@ -0,0 +1,114 @@
+done if the inversion of the loop does not affect program semantics inside
+the loop (order of exceptions thrown etc). The benefit of doing this is the
+fact that there are, in general, instructions that can perform compare/branch
+against zero in an efficient manner and special counter machine register
```
```suggestion
against zero in an efficient manner and special counter machine registers
```

dsouzai commented on doc/compiler/optimizer/LoopOptimizationSummary.md:
```diff
@@ -0,0 +1,114 @@
+Loop Striding creates derived induction variables (e.g. the address calculation
+for array accesses `&a[i]`) and increment/decrement the derived induction
+variables by the appropriate stride through every iteration of the loop.
+This can eliminate some complex array address calculations inside the loop,
+instead replacing by a load (reg or memory). Note that we also create derived
```
```suggestion
This can eliminate some complex array address calculations inside the loop,
replaced instead by a load (reg or memory). Note that we also create derived
```

dsouzai commented on doc/compiler/optimizer/LoopOptimizationSummary.md:
```diff
@@ -0,0 +1,114 @@
+- Check if the loop is known to be short running (either based on induction
+variable information obtained from [Value Propagation](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/ValuePropagation.md))
+or if the special versioning test for short running loops (mentioned earlier)
+is found. In either case, since the loop is provably not long running and does
+not require an async check at all.
```
```suggestion
is found. In either case, since the loop is provably not long running it does
not require an async check at all.
```

dsouzai commented on doc/compiler/optimizer/LoopOptimizationSummary.md:
```diff
@@ -0,0 +1,114 @@
+is found. In either case, since the loop is provably not long running and does
+not require an async check at all.
+- If async checks cannot be skipped completely for the loop, attempt to place
+the sync checks at points within the loop such that there is minimal redundancy
```
```suggestion
the async checks at points within the loop such that there is minimal redundancy
```

dsouzai reviewed (commented):
Thanks for writing this @a7ehuo; it was a very informative read. Mostly minor changes requested.

a7ehuo commented:
>Thanks for writing this @a7ehuo; 

Just to clarify, I didn't write these documents. The original document is written by @vijaysun-omr. I just broke it down into the four individual documents and edited them a bit.

dsouzai commented:
Heh, well thanks for opening the PR notwithstanding :).

hzongaro commented on doc/compiler/optimizer/LoopOptimizationSummary.md:
```diff
@@ -0,0 +1,114 @@
+
+This analysis uses induction variable information found by [Value Propagation](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/ValuePropagation.md).
+Another versioning test that is emitted ensures the loop will not run for a
+very long period of time (in which case the async check inside the loop can
```
Is there any place that describes what an async check?

vijaysun-omr replied:
Not to my knowledge. Async checks get added by the VM at backward branches in the bytecodes on the assumption that one cannot have a loop without a backwards branch. The async check is used for many purposes all of which are some variant of the need to periodically have all application threads yield in order for some sort of action to be taken, with the prime examples being a) GC and b) walk the stack to sample where the thread is executing for driving compilation control. A better definition may be possible, but that hopefully gives the gist.

vijaysun-omr commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+Global Live Variables for GC collects zero initialization information as explained above;
+in addition it also populates each basic block with information about which locals
+are live on entry to the block. This is then used by the code generator to only
+(selectively) switch on bits in the GC stack maps for those reference locals that
```
There is some historical documentation that @0xdaryl had found related to GC stack maps. Maybe he can comment on whether there was some plan to move to the open in the future.

0xdaryl replied:
I believe the material Vijay mentions could be used as a starting point for a GC map overview but more will need to be added to it to make it a useful contribution.  This is on our rather long list of concepts we wish to document as part of the tech debt initiative, but no document exists just yet to link to in this document..

vijaysun-omr commented on doc/compiler/optimizer/ControlFlowOptimizationsSummary.md:
```diff
@@ -0,0 +1,65 @@
+# Control Flow Optimizations
+
+## Block Reordering
+Block Reordering moves blocks around based on making the ‘hot’ path (based on
```
Might be better to name this section `Basic block ordering` to fit in with the actual name of the optimization in the strategy, that is `basicBlockOrdering`

vijaysun-omr commented on doc/compiler/optimizer/ControlFlowOptimizationsSummary.md:
```diff
@@ -0,0 +1,65 @@
+the fall through.
+
+## Block Hoisting
+Block Hoisting duplicates and places the code in a block into its predecessors
+that only have the block as a unique successor, if this would enable better/more
+commoning or other optimizations to occur.
```
I would add one extra sentence : `One example of when this optimization may be effective is with respect to local dead store elimination when the hoisted block ends in a return, since might allow the removal of trailing stores to local variables.`

vijaysun-omr commented on doc/compiler/optimizer/ControlFlowOptimizationsSummary.md:
```diff
@@ -0,0 +1,65 @@
+commoning or other optimizations to occur.
+
+## Block Splitting
+Block Splitting is similar to virtual guard tail splitting except that profiled
```
We may want to consider adding : `The basic idea of block (and virtual guard tail) splitter are the same : namely, do the transformation referred to as "tail duplication" in compiler literature, that essentially involves duplicating code after control flow merge points to avoid the merge in control flow. This transformation should result in more opportunity to specialize and optimize the duplicated code for the hottest paths leading up to the control flow merge.`

vijaysun-omr commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+
+An early pass that splits (duplicates) one allocation into multiple allocations along
+different control flow paths (allocation splitting) is used to maximize the opportunities
+for escape analysis.
```
I agree with @hzongaro There probably is an escape analysis talk and/or presentation from him that we could link to if it is in the open. 

Other points that we probably should mention : `There is a notion of "heapifying" a stack allocated object in cases even when an escape point is detected for the object but the escape point is a cold block. In such cases it is worth keeping the object on stack on the hot path and only if the cold path is taken, then the object is allocated on the heap and the contents of the object on the stack are copied into the newly allocated heap object. As part of heapification, all references from autos/parms to the stack allocated object are updated to point to the heapified object instead. This means that it is safe to execute any code where the object escapes from that point on.`

0xdaryl replied:
EA overview -> https://youtu.be/JMpqwLVwmk8

vijaysun-omr replied:
Other potential additions:

`Our escape analysis implementation also has a notion of de-memoization/re-memoization. This pertains to some special classes in the JCL: Integer, Long, Short, Byte etc. that memoize some instances of those classes in a static array that they maintain (only for some ranges of values). These classes would use these arrays to return an already allocated/memoized object rather than allocate a new object every time (if possible) when `valueOf` is called. These arrays therefore also become escape points for objects that are allocated and stored into them. De-memoization attempts to get around this by converting valueOf calls to new allocations before escape analysis is attempted on the assumption that allocating such objects on the stack would be even more preferable to memoizing them in the arrays. If stack allocation was not possible due to some reason, the de-memoized allocations are re-memoized, leaving the code as it was originally and an attempt is also made to inline the re-memoized valueOf since the inlining of these calls is suppressed in the inliner to give de-memoization a chance later.`

vijaysun-omr replied:
Finally, on the Power platform (maybe AArch64 should use it too if it has a weak memory model, attn : @0xdaryl ) there is a "flow sensitive escape analysis" pass that gets run. 

`Flow sensitive escape analysis tries to find places in the code where an object "has not escaped yet" rather than the notion of "does the object escape anywhere" that is the hallmark of the flow insensitive escape analysis that runs on non-Power platforms. The reason for doing this flow sensitive analysis is because we want to optimize away the lwsync instructions that must be emitted after allocations to ensure that other threads see a properly initialized object. Flow sensitive escape analysis attempts to move such lysyncs together (i.e. combine) as much as possible. This includes trying to move an lwsync for an object to another lwsync (either for the same object (after a constructor if the object has final fields) or a different object), a monexit or volatile store (that do lwsyncs for different reasons) but all only on the basis that the object has not escaped yet.`

vijaysun-omr commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+tries to find explicit definitions of fields (say in constructors) that occur before
+any use of the field. In these cases the zero initialization of fields (required according
+to Java spec, so no field is uninitialized) can be skipped. Note that for reference
+fields the presence of a GC point before any user-specified definition means that
```
The notion of what qualifies as a GC point in the JIT typically depends on the query `node->canGCandReturn()`

a7ehuo replied:
There is a hyperlink link to node->canGCandReturn() in a later section `Live Variables for GC (local and global)`. I'll move that link to the first time when GC point is mentioned. 

```
## Live Variables for GC (local and global)
Live Variables for GC attempts to minimize the number of locals that need to be zero
initialized upon method entry. The basic idea is to perform liveness analysis and find
out if any locals are live at any [GC point](https://github.com/eclipse/omr/blob/81b79405da6c7c960e611a8b2b12fd5861543330/compiler/il/OMRNode.cpp#L2298)
```

vijaysun-omr commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,138 @@
+
+## Global Register Allocator (GRA)
+GRA is a tactical solution for achieving the effect of global register assignment,
+even though the underlying code generators have local register allocators. Basic idea
```
This tactical solution has survived now for over 20 years, i.e. I'd suggest it has earned the right to no longer be referred to as tactical :) @0xdaryl

vijaysun-omr commented on doc/compiler/optimizer/LocalOptimizationsSummary.md:
```diff
@@ -0,0 +1,103 @@
+
+# Local Optimizations
+
+Local optimizations operate on extended basic blocks at a time.
```
Yes, they operate on a single extended basic block at a time.

vijaysun-omr commented on doc/compiler/optimizer/LoopOptimizationSummary.md:
```diff
@@ -0,0 +1,114 @@
+as code does not have to be moved; instead commoning of dominated expressions can be done
+(which is far easier than code motion because of problems introduced by exception checks
+in Java).
+
```
We had a plan to implement loop peeling but we never actually did it. So the documentation should not be included for it.

a7ehuo commented:
@vijaysun-omr @dsouzai @hzongaro 3becd9a addresses all the above comments except for the following two comments. I think they can always be updated once we have an initial version of these documents. It's ready for [another review](https://github.com/eclipse-openj9/openj9/compare/f3755b692379cca72a695e96396d8e03362d09d5..3becd9a659bf3949491c6d694b145d1c75c177a6). Thank! 

1. Add a small description for some pre and post escape analysis passes that were added for handling HCR better https://github.com/eclipse-openj9/openj9/pull/14999#discussion_r869253165
2. Add a link to GC stack map document: https://github.com/eclipse-openj9/openj9/pull/14999#discussion_r868043884

vijaysun-omr commented on doc/compiler/optimizer/GlobalOptimizationsSummary.md:
```diff
@@ -0,0 +1,214 @@
+The reason for doing this Flow Sensitive Escape Analysis is because we want to optimize away the
+`lwsync` instructions that must be emitted after allocations to ensure that other threads
+see a properly initialized object. Flow Sensitive Escape Analysis attempts to move such
+`lysyncs` together (i.e. combine) as much as possible. This includes trying to move an `lwsync`
```
typo : `lysync` -> `lwsync`

vijaysun-omr commented:
Apart from the minor typo I mentioned in my review this morning, I am fine with this version. It can be added to with other optimizations later as a separate work item. Thanks @a7ehuo

hzongaro reviewed (approved):
Looks good - Thanks!

dsouzai commented:
Merging as all reviewers have approved.
